rfcs/src/rfcs/0001-mir-linker.md

```diff
@@ -0,0 +1,244 @@
+- It will enable more powerful and precise static analysis to `kani-compiler`.
+- It won't require any changes to our dependencies.
+- This will fix the harnesses' dependency on the`#[no_mangle]` annotation
+([Issue-661](https://github.com/model-checking/kani/issues/661)).
+
+### Risks
+
+Failures in the linking stage would not impact the tool soundness. I anticipate the following failure scenarios:
+- ICE (Internal compiler error): Some logic is incorrectly implemented and the linking stage crashes.
+  Although this is a bad experience for the user, this will not impact the verification result.
+- Missing items: This would either result in ICE during code generation or a verification failure if the missing
+  item is reachable.
+- Extra items: This shouldn't impact the verification results, and they should be pruned by CBMC's reachability
+  analysis.
+  This is already the case today. In extreme cases, this could include a symbol that we cannot compile and cause an ICE.
+
+The new reachability code would be highly dependent on the `rustc` unstable APIs, which could increase
+the cost of the upstream synchronization.
+That said, the APIs that would be required are already used today.
+
+Finally, this implementation relies on a few unstable options from `cargo` and `rustc`.
+These APIs are used by other tools such as MIRI, so we don't see a high risk that they would be removed.
+
+### Alternatives
+
+The other options explored were:
+1. Pre-compile the standard library and ship the generated `*symtab.json` files.
+2. Pre-compile the standard library, convert them to `goto-programs` (via `symtab2gb`) and link them into
+   one single `goto-program` model. Ship the generated model.
+
+Both alternatives are very similar. They only differ on the artifact that would be shipped.
+They require generating and shipping a custom `sysroot`;
+however, there is no need to implement the reachability algorithm.
+
+We implemented a prototype for the MIR linker and one for the alternatives.
+Both prototypes generate the sysroot as part of the `cargo kani` flow.
+
+We performed a small experiment (on a `c5.4xlarge` ec2 instance running Ubuntu 20.04) to assess the options.
+
+For this experiment, we used the following harness:
+```rust
+#[kani::proof]
+#[kani::unwind(4)]
+pub fn check_format() {
+    assert!("2".parse::<u32>().unwrap() == 2);
+}
+```
+The experiment showed that the MIR linker approach is much more efficient.
+
+See the table bellow for the breakdown of time (in seconds) taken for each major step of
+the harness verification:
+
+
+| Stage                     | MIR Linker | Alternative |
+----------------------------|------------|-------------|
+| compilation               |   22.2s    |    64.7s    |
+| `goto-program` generation |   2.4s     |    90.7s    |
+| `goto-program` linking    |   0.8s     |    33.2s    |
+| code instrumentation      |   0.8s     |    33.1     |
+| verification              |   0.5s     |    8.5s     |
```

Comment:
The results are amazing! I'd expect memory usage to be significantly lower with the MIR linker as well.

If you still have the log files, can you compare the size of the SAT instance (clauses/literals) between the two alternatives? I'm curious to know how much CBMC's slicer is able to reduce the problem size, and how close it can get to the one generated by the MIR linker.

Reply:
Yeah, I totally forgot to include the memory usage here.

Both prototypes run the test described in CI. So you can see the log files for the CI run for [MIR linker](https://pipelines.actions.githubusercontent.com/serviceHosts/14dcd0bf-cf10-416a-af82-d1e50ea29e0a/_apis/pipelines/1/runs/620/signedlogcontent/10?urlExpires=2022-08-31T00%3A11%3A05.1747440Z&urlSigningMethod=HMACV1&urlSignature=5K%2B8bYQk7nidag1e9Ie3uz2aTF7j2OKVXm6oRVQ5C1Y%3D) and the [alternative](https://pipelines.actions.githubusercontent.com/serviceHosts/14dcd0bf-cf10-416a-af82-d1e50ea29e0a/_apis/pipelines/1/runs/622/signedlogcontent/5?urlExpires=2022-08-31T00%3A11%3A10.9310206Z&urlSigningMethod=HMACV1&urlSignature=S7i6ijOKpOYInWy6HTfBT3joP%2B3wQpluhdnDXFBdC1E%3D).

The specific data I believe you are asking for is:
```
14089 variables, 180 clauses
```
and 
```
21610 variables, 7514 clauses
```
for MIR Linker and the alternative, respectively.

It just occurred to me that I should try to check the difference under visualize. It might make it more clear to understand the difference.